docs/DETECTION_RULE_STRUCTURE.md

```diff
@@ -0,0 +1,321 @@
+It contains these three *regex-like* syntax elements:
+- `[ ... ]?` represents an optional builder statement
+- `[ ... ]+` represents one or more repetitions of the enclosed statement
+- `A | B` indicates that exactly one of A or B must be chosen
+
+```java
+new DetectionRuleBuilder<T>()
+     .createDetectionRule()
+     .forObjectTypes(types) | .forObjectExactTypes(types)
+     .forMethods(names) | .forConstructor()
+     [.shouldBeDetectedAs(actionFactory)]?
+     [
+         .withMethodParameter(type) | .withMethodParameterMatchExactType(type)
+         [
+             .shouldBeDetectedAs(valueFactory)
+             [.asChildOfParameterWithId(id)]?
+         ]?
+         [.addDependingDetectionRules(detectionRules)]?
+     ]+
+     .buildForContext(detectionValueContext)
+     .inBundle(bundle)
+     .withDependingDetectionRules(detectionRules) | .withoutDependingDetectionRules()
+```
+
+You may have noticed that this specification does not allow constructing rules for function calls with no parameters.
+Indeed, to keep the specification above simpler, we handle this case separately [later](#special-cases-no-parameters-and-any-parameters).
+
+### Detailed explanations
+
+Writing a detection rule starts with instantiating a new [`DetectionRuleBuilder`](../engine/src/main/java/com/ibm/engine/rule/builder/DetectionRuleBuilder.java) and calling its `createDetectionRule()` method.
+Notice that `DetectionRuleBuilder<T>` is a generic class, that should be parametrized with a language-specific type (learn more [here](./LANGUAGE_SUPPORT.md#identifying-the-four-classes-to-use-in-generics)).
+Then, the two next builder steps aim at identifying the function call that you want to capture, by specifying first its type and then its name.
+
+To specify its type(s), you can use `forObjectTypes(String... types)` to capture any function call that is called on an object whose type is matching one of the provided types **or subtypes**. Instead, you can also use `.forObjectExactTypes(String... types)` to only capture function calls matching one of the provided exact types (and not their subtypes).
+
+> [!NOTE]
+> Some languages, like Python, can have functions defined directly in a file and not in a class, in this case the function is not "called on an object". In this case, the meaning of the provided types depends on the implementation of the language support for the Sonar Cryptography Plugin: you should look into the documentation or code describing what this "object type" represents in your language. In particular, you can look into the implementation of the function `getInvokedObjectTypeString` in the [`ILanguageTranslation`](../engine/src/main/java/com/ibm/engine/language/ILanguageTranslation.java) implementation of your language. For example in Python, we define this type as the "fully qualified name" of the function, which is its full import path.
```

Comment:
@n1ckl0sk0rtge I do think that the meaning of types is given by "the implementation of the language support for the Sonar Cryptography Plugin" and not "the implementation of the language (abstract syntax tree)".
For Python, we _choose_ to define types as the fully qualified name, and it's a choice we implement in the language support layer, rather than something inevitable coming from the AST.

Reply:
```suggestion
> Some languages, like Python, can have functions defined directly in a file and not in a class, in this case the function is not "called on an object". In this case, the meaning of the provided types depends on the implementation of the language support in the engine module: you should look into the documentation or code describing what this "object type" represents in your language. In particular, you can look into the implementation of the function `getInvokedObjectTypeString` in the [`ILanguageTranslation`](../engine/src/main/java/com/ibm/engine/language/ILanguageTranslation.java) implementation of your language. For example in Python, we define this type as the "fully qualified name" of the function, which is its full import path.
```